library/std/src/sys/pal/wasi/fs.rs

```diff
@@ -172,20 +172,22 @@ impl Iterator for ReadDir {
             let offset = if self.offset == self.cap {
                 let cookie = self.cookie.take()?;
                 match self.inner.dir.fd.readdir(&mut self.buf, cookie) {
-                    Ok(bytes) => self.cap = bytes,
+                    Ok(bytes) => {
+                        // No more entries if we read less than buffer size
+                        if bytes < self.buf.len() {
+                            self.cookie = None;
+                            if bytes == 0 {
+                                return None;
+                            }
+                        } else {
+                            self.cookie = Some(cookie);
+                        }
+                        self.cap = self.buf.len();
```

Comment:
I don't think this is correct. In case readdir doesn't fill the entire buffer because there aren't many entries, this will still set the cap to the full buffer, causing us to overflow into the unfilled part and read garbage, since the `data` below will include it.
Or am I misunderstanding this function? It is very confusing, so I wouldn't be surprised.

Reply:
You're right in that `self.cap` should be `bytes`. I think though we would never read unfilled part, as `readdir` will read up to `self.buf.len()`, and unfilled Vec capacity is always filled out with 0s. 
But yeah, this function as a whole if kinda confusing. I've tried to change as little of it as possible.

Reply:
Huh, changing that condition reintroduces the original hang.. O_o